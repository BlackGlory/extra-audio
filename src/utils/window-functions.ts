export type WindowFunction = (alpha: number) => number

export const rectangleWindow: WindowFunction = () => 1

export const triangularWindow: WindowFunction = alpha => {
  return 1 - Math.abs(2 * alpha - 1)
}

export const hannWindow: WindowFunction = alpha => {
  return 0.5 * (1 - Math.cos(2 * Math.PI * alpha))
}

export const hammingWindow: WindowFunction = alpha => {
  return 25 / 46 - (21 / 46) * Math.cos(2 * Math.PI * alpha)
}
