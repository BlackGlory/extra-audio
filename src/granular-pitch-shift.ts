import { clamp } from 'extra-utils'
import { triangularWindow, WindowFunction } from './utils/window-functions.js'
import { getSample } from './utils/get-sample.js'

export function granularPitchShift(
  input: Float32Array
, output: Float32Array
, pitchScale: number
, inputGrainSize: number = 480
, window: WindowFunction = triangularWindow
, overlapRatio: number = 0.5
): Float32Array {
  const outputGrainSize = inputGrainSize / pitchScale

  // 这是假设输出粒子不重叠时的粒子数量.
  // 在`overlapRatio`参数大于0的时候, 输出的粒子会重叠, 实际产生的粒子数量会多于此数值.
  const outputGrains = output.length / outputGrainSize

  const step = 1 - overlapRatio
  for (
    let outputGrainIndex = 0
  ; outputGrainIndex < outputGrains
  ; outputGrainIndex += step
  ) {
    // outputGrainIndex / inputGrainIndex
    // = (outputLength / outputGrainSize) / (inputLength / inputGrainSize)
    // = (1 / outputGrainSize) / (1 / inputGrainSize)
    // = (1 / outputGrainSize) * inputGrainSize
    // = inputGrainSize / outputGrainSize
    // = pitchScale
    const inputGrainIndex = outputGrainIndex / pitchScale

    const inputIndexStart = inputGrainIndex * inputGrainSize
    const outputIndexStart = outputGrainIndex * outputGrainSize

    for (let outputOffset = 0; outputOffset < outputGrainSize; outputOffset++) {
      const inputIndex = inputIndexStart + outputOffset * pitchScale
      // 由于浮点数精度问题, 索引可能会超出`output`的边界.
      const outputIndex = clamp(
        Math.floor(outputIndexStart + outputOffset)
      , [0, output.length - 1]
      )

      output[outputIndex] += getSample(input, inputIndex)
                           * window(outputOffset / (outputGrainSize - 1))
    }
  }

  return output
}
