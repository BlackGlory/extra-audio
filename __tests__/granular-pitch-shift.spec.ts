import { describe, it, expect } from 'vitest'
import { granularPitchShift } from '@src/granular-pitch-shift.js'
import { rectangleWindow } from '@utils/window-functions.js'

describe('granularPitchShift', () => {
  it('pitchScale = 1', () => {
    const input = new Float32Array([0, 1, 2, 3])
    const output = new Float32Array(input.length)
    const pitchScale = 1

    const result = granularPitchShift(
      input
    , output
    , pitchScale
    , 1
    , rectangleWindow
    , 0
    )

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(input)
  })

  it('pitchScale > 1', () => {
    const input = new Float32Array([0, 1, 2, 3])
    const output = new Float32Array(input.length)
    const pitchScale = 2

    const result = granularPitchShift(
      input
    , output
    , pitchScale
    , 1
    , rectangleWindow
    , 0
    )

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(new Float32Array([0.5, 2.5, 4.5, 6]))
  })

  it('pitchScale < 1', () => {
    const input = new Float32Array([0, 1, 2, 3])
    const output = new Float32Array(input.length)
    const pitchScale = 0.5

    const result = granularPitchShift(
      input
    , output
    , pitchScale
    , 1
    , rectangleWindow
    , 0
    )

    expect(result).toBe(output)
    expect(result).not.toBe(input)
    expect(result).toEqual(new Float32Array([0, 0.5, 2, 2.5]))
  })
})
