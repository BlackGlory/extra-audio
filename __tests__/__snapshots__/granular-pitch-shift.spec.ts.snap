// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`granularPitchShift > pitchScale < 1 1`] = `
Float32Array [
  0,
  0.5,
  2,
  2.5,
]
`;

exports[`granularPitchShift > pitchScale > 1 1`] = `
Float32Array [
  0.5,
  2.5,
  4.5,
  6,
]
`;
